File: scc_dag.cpp
std::vector<std::vector<int>> scc_dag(std::vector<std::vector<int>>& scc, std::vector<std::vector<int>>& graph) {
    int n = graph.size();
    int m = scc.size();
    std::vector<int> d(n);
    for(int i = 0; i < m; ++i) for(int v: scc[i]) d[v] = i;
    std::vector<std::vector<int>> to(m);
    std::vector<int> added(m, -1);
    for(int i = 0; i < m; ++i) {
        added[i] = i;
        for(int u: scc[i]) for(int x: graph[u]) {
            int v = d[x];
            if (added[v] != i) {
                added[v] = i;
                to[i].push_back(v);
            }
        }
    }
    return to;
}

File: scc_dag.py
def scc_dag(scc, g):
    n = len(g)
    d = [-1] * n
    for i, c in enumerate(scc):
        for v in c:
            d[v] = i
    m = len(scc)
    to = [[] for _ in range(m)]
    added = [-1] * m
    for i, c in enumerate(scc):
        added[i] = i
        for u in c:
            for v in (d[x] for x in g[u]):
                if added[v] != i:
                    added[v] = i
                    to[i].append(v)
    return to